### PR TITLE
ENH: Extend np.flip to work over multiple axes

### PR DESCRIPTION
**Pull request by bddppq**
#10847 

@mhvk @eric-wieser

### CODE REVIEW

File: numpy/lib/tests/test_function_base.py [hunk 1]
```diff
@@ -107,6 +107,7 @@ def test_axes(self):
         assert_raises(ValueError, np.flip, np.ones(4), axis=1)
         assert_raises(ValueError, np.flip, np.ones((4, 4)), axis=2)
         assert_raises(ValueError, np.flip, np.ones((4, 4)), axis=-3)
+        assert_raises(ValueError, np.flip, np.ones((4, 4)), axis=(0, 3))
```
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py]: All these tests should changed to check for `np.AxisError`

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -157,9 +157,16 @@ def flip(m, axis):
...
         Input array.
-    axis : integer
-        Axis in array, which entries are reversed.
+    axis : None or int or tuple of ints, optional
+         Axis or axes along which to flip over.  The default,
+         axis=None, will flip over all of the axes of the input array.
+         If axis is negative it counts from the last to the first axis.
+
+         If axis is a tuple of ints, flipping is performed on all of the axes
+         specified in the tuple instead of a single axis or all the axes as
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: I don't think "instead of ..." is needed here

File: doc/release/1.15.0-notes.rst [hunk 3]
```diff
@@ -122,6 +122,11 @@ Added experimental support for the 64-bit RISC-V architecture.
 Improvements
 ============
 
+``np.flip`` accepts ``axis`` argument to be None or int or tuple of ints
```
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: Heading would be clearer as ``` ``np.flip`` can operate over multiple axes ```

File: numpy/lib/function_base.py [hunk 4]
```diff
@@ -177,6 +184,10 @@ def flip(m, axis):
     flip(m, 0) is equivalent to flipud(m).
     flip(m, 1) is equivalent to fliplr(m).
     flip(m, n) corresponds to ``m[...,::-1,...]`` with ``::-1`` at position n.
+    flip(m) corresponds to ``m[::-1,::-1,...,::-1]`` with ``::-1`` at all
+    positions.
+    flip(m, (0, 1)) corresponds to ``m[::-1,::-1,...]`` with ``::-1`` at
+    position 0 and position 1.
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: Note that these all render on one line on the web docs, with no line breaks.

File: numpy/lib/function_base.py [hunk 5]
```diff
@@ -199,19 +210,33 @@ def flip(m, axis):
            [[6, 7],
             [4, 5]]])
 
+    >>> np.flip(A)
+    array([[[7, 6],
+            [5, 4]],
+           [[3, 2],
+            [1, 0]]])
+
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: I think these blank lines are breaking the rendering in sphinx

File: numpy/lib/tests/test_function_base.py [hunk 6]
```diff
@@ -173,6 +174,33 @@ def test_4d(self):
...
                          np.flipud(a.swapaxes(0, i)).swapaxes(i, 0))
 
+    def test_default_axis(self):
+        a = np.array([[1, 2, 3],
+                      [4, 5, 6]])
+        b = np.array([[6, 5, 4],
+                      [3, 2, 1]])
+        assert_equal(np.flip(a), b)
+
+    def test_multiple_axes(self):
```
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py]: Add a test for `axis=()`, which does no flips?

**Comment by bddppq** [comment]
@eric-wieser Thanks for your quick review. I have addressed your comments, please take another look at your convenience.

File: doc/release/1.15.0-notes.rst [hunk 7]
```diff
@@ -122,6 +122,11 @@ Added experimental support for the 64-bit RISC-V architecture.
 Improvements
 ============
 
+``np.flip`` can operate over multiple axes
+------------------------------------------------------------------------
```
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: I think sphinx gets unhappy if this underline is too long like this

**Review by eric-wieser — APPROVED** [review]

**Comment by mhvk** [comment]
This looks good! One last request: could you squash all the commits into one, and start the title of that single commit with `ENH:`? See https://docs.scipy.org/doc/numpy-dev/dev/gitwash/development_workflow.html#writing-the-commit-message

**Comment by bddppq** [comment]
@mhvk Sure, I should have read the guideline before making the PR :-)

File: doc/release/1.15.0-notes.rst [hunk 8]
```diff
@@ -122,6 +122,11 @@ Added experimental support for the 64-bit RISC-V architecture.
 Improvements
 ============
 
+``np.flip`` can operate over multiple axes
+------------------------------------------
+``np.flip`` now accepts ``axis`` argument to be None or int or tuple of ints.
```
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: nit: Grammar of this line would be better as ``` ``np.flip`` now accepts tuples of int and None in its ``axis`` argument ```

File: numpy/lib/function_base.py [hunk 9]
```diff
@@ -157,9 +157,15 @@ def flip(m, axis):
     ----------
     m : array_like
         Input array.
-    axis : integer
-        Axis in array, which entries are reversed.
+    axis : None or int or tuple of ints, optional
```
Comment (hameerabbasi) [inline numpy/lib/function_base.py]: Maybe rewrite this to be `axis : {None, int, Tuple[int]}`?

**Review by hameerabbasi — COMMENTED** [review]
Minor change in docs.

**Comment by bddppq** [comment]
the "lgtm" failure doesn't seem related

**Comment by hameerabbasi** [comment]
@bddppq Did you possibly do a merge commit instead of a rebase?

**Comment by mattip** [comment]
lgtm seems flakey on documentation-only pull requests, I have [asked them](https://discuss.lgtm.com/t/yet-another-commits-either-could-not-be-fetched-or-could-not-be-merged/959) what is going on

File: numpy/lib/function_base.py (hunk 9 above)
Comment (eric-wieser) [reply numpy/lib/function_base.py]: @hameerabbasi: Do we already write it like that elsewhere?

**Comment by eric-wieser** [comment]
@hameerabbasi: Looks like a rebase to me.

LGTM seems to be back on it's feet, so this LGTM too. I'll leave it for @mhvk to merge, since he proposed a different implementation to the one I proposed.

Thanks for taking the time to write the tests and docs!

File: numpy/lib/function_base.py (hunk 9 above)
Comment (hameerabbasi) [reply numpy/lib/function_base.py]: @eric-wieser Ah, I thought we did, apparently we don't. It's like this in `np.sum`, for example.

**Review by hameerabbasi — APPROVED** [review]
Looks good to me. +1.

File: numpy/lib/function_base.py [hunk 10]
```diff
@@ -157,9 +157,15 @@ def flip(m, axis):
...
+         Axis or axes along which to flip over. The default,
+         axis=None, will flip over all of the axes of the input array.
+         If axis is negative it counts from the last to the first axis.
+
+         If axis is a tuple of ints, flipping is performed on all of the axes
+         specified in the tuple.
 
+         .. versionadded:: 1.15.0
```
Comment (mhvk) [inline numpy/lib/function_base.py]: It is a detail, but this should be 
```
.. versionchanged:: 1.15.0
   None and tuples of axes are supported
```

**Review by mhvk — COMMENTED** [review]
Looks all OK and I'd be happy to merge except for the nit about the docstring: since `axis` already existed as an argument, I think it has to be `versionchanged`.

File: doc/release/1.15.0-notes.rst [hunk 11]
```diff
@@ -122,6 +122,11 @@ Added experimental support for the 64-bit RISC-V architecture.
 Improvements
 ============
 
+``np.flip`` can operate over multiple axes
+------------------------------------------
+``np.flip`` now accepts tuples of int and None in its ``axis`` argument. If
```
Comment (stefanv) [inline doc/release/1.15.0-notes.rst]: This is very easily misread as "tuples-of-in-and-None".  Perhaps "accepts None, or tuples of int, in its ... "

**Review by stefanv — APPROVED** [review]

**Comment by mhvk** [comment]
@bddppq - it does seem we're getting at you slowly, one by one, rather than just let you do the changes and be done with it! Anyway, if you have the patience, could you address @stefanv's comment as well?

**Comment by stefanv** [comment]
Yes, sorry about that; what's the NumPy policy about pushing back to contributed PRs?  I could just make this change myself.

**Comment by bddppq** [comment]
@mhvk @stefanv No worry I just updated the doc. Doc clarity is very important for such a widely used library, so I'm happy to polish it :-)

**Comment by mhvk** [comment]
@stefanv - good question... My main hesitation is that we try to also keep commits self-contained (which I really like), and to make a small change, plus rebase/squash is quite unhandy. @charris: any advice on the extent to which we should just make small textual edits?

Anyway, off topic otherwise: @bddppq: thanks for your contribution! Will merge this once tests have passed.

**Comment by stefanv** [comment]
FWIW, on scikit-image we now allow reviewers pushing back small fixes to PRs, and also allow authors to make multiple commits, and even merge with master (easier than rebase, quite often).  The key then is that we do a squash+merge, editing the commit message as appropriate.

**Comment by mhvk** [comment]
Thanks, @bddppq!

**Comment by mhvk** [comment]
@stefanv - squash + merge via github? I vaguely recall people being unhappy with this, but do not really remember why.

**Comment by stefanv** [comment]
@mhvk I guess I was one of those people, when I still believed it was better to prepare "pristine patch sets".  Now, I'm quite happy to make contribution work in whichever way we can.

**Comment by eric-wieser** [comment]
Non-pristine patch sets tend to correlate with things that only really need one commit anyway, so I tend to fall in favor of doing a squash and merge in these cases.

**Pull request merged**